### PR TITLE
Bug 1201062 - port slugid node.js project to python

### PR DESCRIPTION
**Pull request by petemoore**
I've pre-emptively [published to pypi](https://pypi.python.org/pypi?:action=display&name=slugid&version=1.0.3). It took a couple of bumps to get it right, hence version 1.0.3.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+Slugs are url-safe base64 encoded v4 uuids, stripped of base64 ``=`` padding.
+
+There are two methods for generating slugs - ``slugid.v4()`` and
+``slugid.nice()``.
+
+The ``slugid.v4()`` method returns a slug from a randomly generated v4 uuid.
+The ``slugid.nice()`` method returns a v4 slug which conforms to a set of
+"nice" properties. At the moment the only "nice" property is that the slug
+starts with [A-Za-f], which in turn implies that the first (most significant)
+but of its associated uuid is set to 0.
```
Comment (djmitche) [inline README.rst]: bit

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+Generated slugs take the form ``[A-Za-z0-9_-]{22}``, or more precisely:
+
+- ``slugid.v4()`` slugs conform to
+  ``[A-Za-z0-9_-]{8}[Q-T][A-Za-z0-9_-][CGKOSWaeimquy26-][A-Za-z0-9_-]{10}[AQgw]``
+
+- ``slugid.nice()`` slugs conform to
+  ``[A-Za-f][A-Za-z0-9_-]{7}[Q-T][A-Za-z0-9_-][CGKOSWaeimquy26-][A-Za-z0-9_-]{10}[AQgw]``
+
+RFC 4122 defines the setting of 6 bits of the v4 UUID which determines these
+regular expressions.
```
Comment (djmitche) [inline README.rst]: Might want to add that this is why there are 122 bits of entropy in a 128-bit uuid (since I had forgotten)

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+complete license.
+
+Publishing
+----------
+To republish this library to pypi.python.org, update the version number in
+``slugid/__init__.py``, commit it, push to github, and then run:
+
+.. code-block:: python
+
+    python setup.py sdist upload --sign
```
Comment (djmitche) [inline README.rst]: Really tangential, but you should consider using `twine` which allows signing during upload

**Comment by djmitche** [comment]
You might also mention that the rng characteristics match those of the built-in `uuid` library.  That library doesn't document its randomness, but falls back to system uuid-generation libraries where available, then urandom, then random -- so it's as good as it can get on the given system.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    description='Base64 encoded uuid v4 slugs',
+    author='Pete Moore',
+    author_email='pmoore@mozilla.com',
+    url='http://taskcluster.github.io/slugid.py',
+    packages=packages,
+    package_data={'': ['LICENSE', 'README.md']},
+    license='MPL 2.0',
+    classifiers=(
+        'Intended Audience :: Developers',
+        'Natural Language :: English',
```
Comment (walac) [inline setup.py:36]: nit: Add supported python versions, like [here](https://github.com/walac/pyusb/blob/master/setup.py#L71-L75)

File: slugid/__init__.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+   >>> slugid.v4()
+   -9OpXaCORAaFh4sJRk7PUA
+"""
+
+__title__ = 'slugid'
+__version__ = '1.0.3'
+__author__ = 'Peter Moore'
+__license__ = 'MPL 2.0'
+
+from .slugid import decode, encode, nice, v4
```
Comment (walac) [inline slugid/__init__.py:43]: If there is only one module, any special reason why it should be a package?

File: slugid/slugid.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+import uuid
+import base64
+
+def encode(uuid_):
+    """
+    Returns the given uuid as a 22 character slug. This can be a regular v4
+    slug or a "nice" slug.
+    """
+    bytes = uuid_.bytes
```
Comment (walac) [inline slugid/slugid.py]: [bytes](https://docs.python.org/3/library/functions.html#bytes) is currently a built-in type in Python, may we call it another name?

File: slugid/slugid.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    slug = base64.urlsafe_b64encode(str(bytes))[:22] # Drop '==' padding
```
Comment (walac) [inline slugid/slugid.py]: Do we want the first 22 characters or remove the last two? If it is the second option, I rather prefer the `[:-2]` idiom.

edit: or `.replace("==", "")`

File: test.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    set) to its known slug. The specific uuid was chosen since it has a slug
+    which contains both `-` and `_` characters."""
+
+    # 10000000010011110011111111001000110111111100101101001011000001101000100111111011101011101111101011010101111000011000011101010100....
+    # <8 ><0 ><4 ><f ><3 ><f ><c ><8 ><d ><f ><c ><b ><4 ><b ><0 ><6 ><8 ><9 ><f ><b ><a ><e ><f ><a ><d ><5 ><e ><1 ><8 ><7 ><5 ><4 >
+    # < g  >< E  >< 8  >< _  >< y  >< N  >< _  >< L  >< S  >< w  >< a  >< J  >< -  >< 6  >< 7  >< 6  >< 1  >< e  >< G  >< H  >< V  >< A  >
+    uuid_ = uuid.UUID('{804f3fc8-dfcb-4b06-89fb-aefad5e18754}')
+    expectedSlug = 'gE8_yN_LSwaJ-6761eGHVA'
+
+    # Encode
```
Comment (walac) [inline test.py]: I think this comment is verbose, since the code is self explanatory.

File: test.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    actualSlug = slugid.encode(uuid_)
+
+    # Test that it encoded correctly
```
Comment (walac) [inline test.py]: ditto.

Same applies for `testDecode`, `testUuidEncodeDecode`, `testSlugDecodeEncode`.

File: test.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    # Each entry is a dict with a key for each character found, and its value
+    # as the number of times that character appeared at that position in the
+    # slugid in the large sample of slugids generated in this test.
+    k = [{}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}]
+
+    # Generate a large sample of slugids, and record what characters appeared
+    # where...  A monte-carlo test has demonstrated that with 64 * 20
+    # iterations, no failure occurred in 1000 simulations, so 64 * 40 should be
+    # suitably large to rule out false positives.
+    for i in range(0, 64 * 40):
```
Comment (walac) [inline test.py:144]: ~~If we are sticking with Python 2, might be better use [xrange](https://docs.python.org/2/library/functions.html#xrange).~~

edit: I am not sure if a 2560 elements list is such a disaster worth to use a deprecated function.

File: test.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                k[j][slug[j]] = k[j][slug[j]] + 1
+            else:
+                k[j][slug[j]] = 1
+
+    # Compose results into an array `actual`, for comparison with `expected`
+    actual = []
+    for j in range(0, len(k)):
+        actual.append('')
+        for a in k[j].keys():
+            if k[j][a] > 0:
```
Comment (walac) [inline test.py:158]: iiuc, if `k[j][a] == 0` then `k[j][a]` will raise [KeyError](https://wiki.python.org/moin/KeyError), right? Doesn't it make more sense `k` be an array of [set](https://docs.python.org/2/library/stdtypes.html#set)s? So the test would be:

``` python
if a in k[j]:
```

File: test.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    if a == b:
+        return True
+    if a is None or b is None:
+        return False
+    if len(a) != len(b):
+        return False
+    for i in range (0, len(a)):
+        if a[i] != b[i]:
+            return False
+    return True
```
Comment (walac) [inline test.py]: The function can be implemented simpler:

``` python
def arraysEqual(a, b):
    return cmp(a, b) == 0
```

**Comment by petemoore** [comment]
Thanks guys for all your awesome comments! I'll work on these in the morning. Really appreciated. :+1:

File: test.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    actualUuid = slugid.decode(slug)
+  
+    # Test that it is decoded correctly
+    assert expectedUuid == actualUuid, "Slug not correctly decoded into uuid: '" + str(expectedUuid) + "' != '" + str(actualUuid) + "'"
+  
+
+def testUuidEncodeDecode():
+    """ Test that 100 v4 uuids are unchanged after encoding and then decoding them"""
+
+    for i in range(0, 100):
```
Comment (walac) [inline test.py]: After some chat in irc, we are going to run this test 10,000 times.

File: test.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        slug = slugid.encode(uuid_)
+
+        # Test that decode uuid matches original
+        assert slugid.decode(slug) == uuid_, "Encode and decode isn't identity"
+
+
+def testSlugDecodeEncode():
+    """ Test that 100 v4 slugs are unchanged after decoding and then encoding them."""
+
+    for i in range(0, 100):
```
Comment (walac) [inline test.py]: Same here.

File: test.py (hunk 11 above)
Comment (petemoore) [inline test.py:158]: Originally I had it as a set, but then it occurred to me that if this test should ever fail, it is useful to have the spread data - i.e. to be able to see in a large sample, how many times a particular character appeared - this way you immediately get a good visualisation if the distribution is even (i.e. if the letter 'A' appears 300 times but the number '5' appears 4 times there is no doubt something wrong). It is true the logic is simpler with a set, but I think the benefit of keeping the total is that if it fails, we have much stronger data to understand why.

File: test.py (hunk 12 above)
Comment (petemoore) [inline test.py]: This is a nice shorter version, thanks!
I spotted `cmp` built in function got removed in python 3, but if we choose to support python 3 later, there is the [cmp_to_key](https://docs.python.org/3/library/functools.html#functools.cmp_to_key) that should help us migrate.

File: test.py (hunk 10 above)
Comment (petemoore) [inline test.py:144]: OK, I'll leave it as it is then. Agreed that it is quite a small range so hopefully should be ok.

File: test.py (hunk 14 above)
Comment (petemoore) [inline test.py]: Will do.

File: test.py (hunk 13 above)
Comment (petemoore) [inline test.py]: Will do.

File: test.py (hunk 8 above)
Comment (petemoore) [inline test.py]: I'll remove them.

File: test.py (hunk 9 above)
Comment (petemoore) [inline test.py]: OK.

File: slugid/slugid.py (hunk 7 above)
Comment (petemoore) [inline slugid/slugid.py]: OK

File: slugid/slugid.py (hunk 6 above)
Comment (petemoore) [inline slugid/slugid.py]: Will remove the variable completely where possible, and call `rawBytes` elsewhere.

File: setup.py (hunk 4 above)
Comment (petemoore) [inline setup.py:36]: Good spot - will do. Thanks!

File: slugid/__init__.py (hunk 5 above)
Comment (petemoore) [inline slugid/__init__.py:43]: Just in anticipation of possible future expansion... Might be overkill.

**Pull request merged**